File: src/main/java/spring/springserver/domain/auth/service/AuthService.java
package spring.springserver.domain.auth.service;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import spring.springserver.domain.auth.data.request.GenerateTokenRequest;
import spring.springserver.domain.auth.data.request.SignInRequest;
import spring.springserver.domain.auth.data.request.SignUpRequest;
import spring.springserver.domain.auth.data.response.SignInResponse;
import spring.springserver.domain.auth.data.response.SignUpResponse;
import spring.springserver.domain.auth.data.response.SignOutResponse;
import spring.springserver.domain.auth.exception.AuthStatusCode;
import spring.springserver.domain.member.entity.Member;
import spring.springserver.domain.member.repository.MemberRepository;
import spring.springserver.global.exception.exception.ApplicationException;

import java.util.Optional;

@Service
@RequiredArgsConstructor
@Transactional(rollbackFor = Exception.class)
public class AuthService {

    private final PasswordEncoder passwordEncoder;
    private final MemberRepository memberRepository;
    private final TokenService tokenService;

    /**
     * 회원가입
     * @param signUpRequest 회원가입 요청 값
     * @return message
     */
    public SignUpResponse signUp(SignUpRequest signUpRequest) {

        if (memberRepository.existsByUsername(signUpRequest.username())) {

            throw new ApplicationException(AuthStatusCode.USERNAME_ALREADY_EXIST);
        }

        if (memberRepository.existsByEmail(signUpRequest.email())) {

            throw new ApplicationException(AuthStatusCode.EMAIL_ALREADY_EXIST);
        }

        if (memberRepository.existsByPhone(signUpRequest.phone())) {

            throw new ApplicationException(AuthStatusCode.PHONE_ALREADY_EXIST);
        }

        memberRepository.save(signUpRequest.toEntity(passwordEncoder.encode(signUpRequest.password())));

        return SignUpResponse.of("회원가입이 완료되었습니다.");
    }

    /**
     * @param signInRequest 로그인 요청 값
     * @param httpServletResponse 쿠키 저장용
     * @return accessToken, refreshToken
     */
    @Transactional(readOnly = true)
    public SignInResponse signIn(SignInRequest signInRequest,
                                 HttpServletResponse httpServletResponse) {

        Member member = Optional.ofNullable(memberRepository.findByUsername(signInRequest.username()))
                .orElseThrow(
                        () -> new ApplicationException(AuthStatusCode.INVALID_CREDENTIALS)
                );

        if (!passwordEncoder.matches(signInRequest.password(), member.getPassword())) {

            throw new ApplicationException(AuthStatusCode.INVALID_CREDENTIALS);
        }

        GenerateTokenRequest generateTokenRequest = new GenerateTokenRequest (
            member.getUsername(),
            member.getRole()
        );

        return SignInResponse.of(
                tokenService.generateAccessToken(
                        generateTokenRequest,
                        httpServletResponse
                ),
                tokenService.generateRefreshToken(
                        generateTokenRequest,
                        httpServletResponse
                )
        );
    }

    /**
     *
     * @param httpServletRequest 쿠키용
     * @param httpServletResponse 쿠키용
     * @return message
     */
    public SignOutResponse signOut(HttpServletRequest httpServletRequest,
                                   HttpServletResponse httpServletResponse) {

        tokenService.deleteTokens(
                httpServletRequest,
                httpServletResponse
        );

        return SignOutResponse.of("로그아웃 되었습니다.");
    }
}

File: src/main/java/spring/springserver/domain/auth/service/auth/AuthService.kt
package spring.springserver.domain.auth.service.auth

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import spring.springserver.domain.auth.data.request.GenerateTokenRequest
import spring.springserver.domain.auth.data.request.OAuthExchangeRequest
import spring.springserver.domain.auth.data.request.SignInRequest
import spring.springserver.domain.auth.data.request.SignUpRequest
import spring.springserver.domain.auth.data.response.SignInResponse
import spring.springserver.domain.auth.data.response.SignOutResponse
import spring.springserver.domain.auth.data.response.SignUpResponse

interface AuthService {

    fun signUp(
        signUpRequest: SignUpRequest
    ) : SignUpResponse

    fun signIn(
        signInRequest: SignInRequest,
        httpServletResponse: HttpServletResponse
    ) : SignInResponse

    fun signOut(
        httpServletRequest: HttpServletRequest,
        httpServletResponse: HttpServletResponse
    ) : SignOutResponse

    /**
     * 토큰 쿠키가 닿지 않는 origin으로 소셜 로그인을 마칠 때 토큰 대신 넘기는 1회용 코드.
     */
    fun issueOAuthExchangeCode(
        generateTokenRequest: GenerateTokenRequest
    ): String

    fun exchangeOAuthCode(
        oAuthExchangeRequest: OAuthExchangeRequest,
        httpServletResponse: HttpServletResponse
    ): SignInResponse

    fun verifyPassword(
        httpServletRequest: HttpServletRequest,
        rawPassword: String
    ): Boolean
}